book-content/chapters/13-modules-scripts-declaration-files.md

```diff
@@ -0,0 +1,988 @@
+
+#### Inside a Script, `declare module` Overrides
+
+Let's go back to our `express.d.ts` file. If we remove the `export {}` statement, it will be treated as a script:
+
+```typescript
+// inside express.d.ts
+
+declare module "express" {
+  export interface MyType {
+    hello: string;
+  }
+}
+```
+
+Now, we've completely overridden the `express` module. This means that the `express` module no longer has any exports except for `MyType`:
+
+```ts
+// anywhere.ts
+import { Express } from "express"; // red squiggly line under "Express"
+```
+
+Just like module augmentation, we can get the same behavior by changing `express.d.ts` to `express.ts` (if `moduleDetection` is set to `auto`).
+
+So, just the presence or absence of an `export` statement can radically change the behavior of `declare module`.
+
+Overriding is occasionally useful when you want to completely replace the types of a module, perhaps when a third-party library has incorrect types.
+
+## Declaration Files You Don't Control
+
+You might think that declaration files are a relatively niche feature of TypeScript. But in every project you create, you're likely using hundreds of declaration files. They either ship with libraries, or come bundled with TypeScript itself.
+
+### TypeScript's Types
+
+Whenever you use TypeScript, you're also using JavaScript. JavaScript has many built-in constants, functions and objects that TypeScript needs to know about. A classic example are the array methods.
+
+```ts
+const numbers = [1, 2, 3];
+
+numbers.map((n) => n * 2);
+```
+
+Let's step back for a minute. How does TypeScript know that `.map` exists on an array? How does it know that `.map` exists, but `.transform` doesn't? Where is this defined
+
+As it turns out, TypeScript ships with a bunch of declaration files that describe the JavaScript environment. We can do a 'go to definition' on `.map` to see where that is:
+
+```ts
+// inside lib.es5.d.ts
+
+interface Array<T> {
+  // ... other methods ...
+  map<U>(
+    callbackfn: (value: T, index: number, array: T[]) => U,
+    thisArg?: any,
+  ): U[];
+}
+```
+
+We've ended up in a file called `lib.es5.d.ts`. This file is part of TypeScript, and describes what JavaScript looked like in ES5, a version of JavaScript from 2009. This is when `.map` was introduced to JavaScript.
+
+Another example would be `.replaceAll` on strings:
+
+```ts
+const str = "hello world";
+
+str.replaceAll("hello", "goodbye");
+```
+
+Doing a 'go to definition' on `.replaceAll` will take you to a file called `lib.es2021.string.d.ts`. This file describes the string methods that were introduced in ES2021.
+
+Looking at the code in `node_modules/typescript/lib`, you'll see dozens of declaration files that describe the JavaScript environment.
+
+Understanding how to navigate these declaration files can be very useful for fixing type errors. Take a few minutes to explore what's in `lib.es5.d.ts` by using 'go to definition' to navigate around.
```

Comment:
For funsies... one could swap out all lib files and use a custom module describing even how an `Array` interface should look like and more. There are options to do that :D I'd really expect that there would be dragons if you'd do that for some very core functionality though. Although internally, TS always tries to load those definitions from the used lib files.